bcda/models/models.go

```diff
@@ -40,14 +45,93 @@ func InitializeGormModels() *gorm.DB {
 type Job struct {
 	gorm.Model
 	ACO        ACO       `gorm:"foreignkey:ACOID;association_foreignkey:UUID"` // aco
-	ACOID      uuid.UUID `gorm:"primary_key; type:char(36)" json:"aco_id"`
+	ACOID      uuid.UUID `gorm:"type:char(36)" json:"aco_id"`
 	User       User      `gorm:"foreignkey:UserID;association_foreignkey:UUID"` // user
 	UserID     uuid.UUID `gorm:"type:char(36)"`
 	RequestURL string    `json:"request_url"` // request_url
 	Status     string    `json:"status"`      // status
+	JobCount   int
 	JobKeys    []JobKey
 }
 
+func (job *Job) CheckCompleted() (bool, error) {
+
+	// Trivial case, no need to keep going
+	if job.Status == "Completed" {
+		return true, nil
+	}
+	db := database.GetGORMDbConnection()
+	defer database.Close(db)
+
+	var completedJobs int64
+	db.Model(&JobKey{}).Where("job_id = ?", job.ID).Count(&completedJobs)
+
+	if int(completedJobs) >= job.JobCount {
+
+		staging := fmt.Sprintf("%s/%d", os.Getenv("FHIR_STAGING_DIR"), job.ID)
+		err := os.Remove(staging)
+
+		if err != nil {
+			log.Error(err)
+		}
+
+		return true, db.Model(&job).Update("status", "Completed").Error
+
+	}
+
+	return false, nil
+}
+
+func (job *Job) GetEnqueJobs(encrypt bool, t string) (enqueJobs []*que.Job, err error) {
+	var jobIDs []string
+
+	var rowCount = 0
+	maxBeneficiaries := utils.GetEnvInt("BCDA_FHIR_MAX_RECORDS", BCDA_FHIR_MAX_RECORDS_DEFAULT)
```

Comment:
Am I following correctly, that you assume (& enforce) that each ACO will have <= 10K beneficiaries?

If so, why?

Reply:
I do not.  I enforce that I will que up a new internal job when I get over 10K beneficiaries.  If an ACO has 55,000 benes I will que up 5 jobs with 10K benes and one more with 5,000 benes.  This is our attempt to parallelize things while not taking Blue Button down.  The que has a max simultaneous workers which will be limited by how many connections BB can handle simultaneously.

Reply:
Cool!

Some questions:

1. Does each job just run a single request to the backend at a time, or does it split out the requests over goroutines or some such?
2. What's the current default for max concurrent workers?
3. Are workers 1:1 with EC2 instances, or something else?

Reply:
1. This seemed like an easier task than doing go routines because it achieves the same end result with a lot less work by leveraging an existing worker queue.  each Queue job will run through the bene list sequentially until it completes.  When all the Queue jobs are done the Job is done and ready to be downloaded by the end user.
2. I think it is 2, but the expectation is that it will be much higher than that in production.  there is a var called "WORKER_POOL_SIZE" that controls it
3. There should be many workers per EC2 instance.

Reply:
What job primitive are workers run in then? Separate process per worker? Separate thread? Separate goroutine?

Reply:
The workers are running using QueueGO which is a Go port of Ruby Queue.  I don't have a great understanding of the internals, I just know it queues jobs, limits how many jobs can run simultaneously and won't fall apart if the system goes down.  


    How many EC2 instances do I have to run workers on? 
       Not sure.  I think this isn't a static value and is subject to auto scaling.  I think each EC2 instance has a fixed number of workers.
    How many user jobs/requests will I allow to be worked at once?
    a Job is what our end user experiences, a queueJob is what I am adding.  Each worker thread will run one queueJob at a time.  This could be one Job or a different job for each worker.  The queue works a queueJob until it is done and then grabs another one.  The size of the worker pool limits how many queueJobs will be worked simultaneously which limits the load on BB.
    
If payloads are split into multiple files/chunks, what is the max size of those chunks?

one queueJob makes one file.  By limiting the number of benes the file size is inherently limited and we don't have any concurrency issues with file writes.  It would be theoretically possible for the 10,000 benes with the most claims to all end up in the same file chunk and the file would be rather large, but I feel that at 10,000 the law of averages has to kick in and file sizes will be within a small range of each other.  To tune it we can alter the number of benes per file if the file sizes are too big or too small.
    How many requests to the backend Data Server will those EC2 instances collectively be running concurrently?  This is the number of worker threads.  We can increase or decrease this number easily as needs demand.

If you want to have a call or a meeting on this I am happy to organize one.

Reply:
I think ultimately the management is actually pretty simple and needs to satisfy only 2 constraints; how many benes can I put in a file before it starts to become "too big" (limits of encryption and file download times are limiting factors) and how many simultaneous calls to BB can I make before I break it.  

The max bene parameter will control how big the files get because barring some unexpected statistical bias 10,000 benes in a file be approximately 10,000 x the average file size per bene.  The goal is not to even come in the neighborhood of the limits of encryption file size but to have manageable sized files for our users to get them in a reasonable amount of time and not have to ever restart a 1 hour download from scratch.  10,000 is definitely just a guess, but it puts an upper limit on the number of files at around 20 per ACO (biggest ACO is +-200,000 benes afaik).  If these files are all pushing 64 GB we have a much different problem than any of us realize.  

The max requests to BB is determined by the size of the worker pool.  If we have 100 workers that is 100 max connections to BB; if we have 10,000 workers that is 10,000 connections.  If we can have 10,000 connections at a time, but only have 100 queuejobs cooking at a time we are not running in an optimized way, but this is going to be an order of magnitude faster than what we have now and has the added benefit of not writing a ton of complicated threading code ourselves.

Reply:
Couple follow-up questions:

1. What controls how many worker EC2 instances are running?
2. Is `WORKER_POOL_SIZE` applied per worker EC2 instance, or globally?

I'm wondering if autoscaling is going to come in Kool-Aid Man style and suddenly double or triple the `WORKER_POOL_SIZE`, sending way more connections to the backend than expected.

Reply:
Hi Karl,
1) For the worker EC2 instances, we currently have it configured in [this manner](https://github.com/CMSgov/bcda-ops/blob/master/terraform/opensbx/main.tf#L646-L648) (2 desired, 2 min, 2 max).  We scale up and down based on various auto-scaling policies defined [here](https://github.com/CMSgov/bcda-ops/blob/master/terraform/modules/asg/main.tf#L120).  Obviously these are tweakable as we learn more about handling of load.
2) `WORKER_POOL_SIZE` is applied globally across each EC2 instance that spins up. The default is 2 (for now), but obviously tweakable.

Reply:
I think it is 4 max, but with the worker pool limited across instance I don't think it will matter too much because spinning up another instance won't increase the number of running tasks.   I think that brings up a different point of what auto scaling will accomplish and what problems it will solve, but that is a discussion for another time.

Reply:
I made a tech spec back in December for this and updated it recently with a new diagram if you want to take a look at that it might answer some questions.

https://confluence.cms.gov/display/BCDA/Tech+Spec%3A+Worker

Reply:
> WORKER_POOL_SIZE is applied globally across each EC2 instance that spins up. The default is 2 (for now), but obviously tweakable.

If it's constrained globally, that covers my main concern. Out of curiosity, what's enforcing/synchronizing that constraint globally?

Reply:
@karlmdavis To clarify, the global definition of `WORKER_POOL_SIZE` is the value that each EC2 instance will inherit.  So, if our `WORKER_POOL_SIZE` is set to 2, then each EC2 instance that spins up will have a worker pool size of 2.

Reply:
> if our WORKER_POOL_SIZE is set to 2, then each EC2 instance that spins up will have a worker pool size of 2

Okay, then we're back to me being a bit concerned. Some background:

* In order to produce data in a timely fashion, BCDA will likely want to have hundreds of concurrent requests going. Certainly 150, and I would not be at all surprised to see that eventually scale up to 1000.
* Each of those requests is super lightweight: if we can support serving it with just a couple of servers (and we can) then you definitely shouldn't need more hardware than that as a client.
* During autoscaling events (e.g. deploys, health check failures, scaling up if you ever add a policy for that, etc.) your number of workers could easily go to 150% of baseline, and perhaps sometimes 200%.

That strikes me as a risky design. Given that it's your design, it presumably doesn't seem like a bad idea to you, though. Can you help me understand why you don't find that a bit concerning?

Reply:
Hi @karlmdavis, kindly requesting that we move the design discussion to this Tech Spec:
https://confluence.cms.gov/display/BCDA/Tech+Spec%3A+Worker

I have populated the comments section of the doc above with some information that might be helpful.